Move continuous integration from Travis CI to Github Actions

IroNEDR opened the pull request:
Our free travis CI credit ran out a long time ago. With this PR we move the our current CI Pipeline from travis to github actions.

With this new pipeline we test for Python versions 3.6 - 3.12. During the process some regressions in our unit tests on older Python versions were discovered such as using the `tuple` instead of `Tuple` in the type annotations. These small fixes are included in this PR.

An interesting read regarding costs etc.:
https://docs.github.com/en/actions/administering-github-actions/usage-limits-billing-and-administration

According to this, we fall under the free category.

vc1492a reviewed (approved):
Looks great @IroNEDR, thanks for the work! ✨

vc1492a commented:
@IroNEDR I have approved the PR, looks good to me. I will hold off on merging and deleting the feature branch for now, but next time if you would like me to proceed with the merge and with deleting the feature branch please let me know 🔧

IroNEDR commented:
@vc1492a no worries, good thing you didn't merge yet. I renamed the workflow file and added a test status badge to the readme to replace the old travis badge
